granary/reddit.py

```diff
@@ -0,0 +1,215 @@
+# coding=utf-8
+"""reddit.com source class.
+"""
+
+from . import source
+import datetime
+import logging
+from oauth_dropins import reddit
+from oauth_dropins.webutil import util
+import re
+import urllib.parse, urllib.request
+
+import praw
+
+class Reddit(source.Source):
+
+    DOMAIN = 'reddit.com'
+    BASE_URL = 'https://reddit.com'
+    NAME = 'reddit'
+
+    def __init__(self, refresh_token):
+        self.refresh_token = refresh_token
+
+    @staticmethod
+    def timestamp_to_iso8601(time_stmp):
+        """Converts a timestamp from UNIX format to ISO 8601.
+        """
+        if not time_stmp:
+            return None
+
+        return datetime.datetime.fromtimestamp(time_stmp).isoformat()
+
+    def redditor_to_actor(self, user):
+        """Converts a praw Redditor to an actor.
+        https://praw.readthedocs.io/en/latest/code_overview/models/redditor.html
+
+        Args:
+          user: praw Redditor object
+
+        Returns:
+          an ActivityStreams actor dict, ready to be JSON-encoded
+        """
+        username = user.name
+        if not username:
+            return {}
+
+
+        # trying my best to grab all the urls from the profile description
+
+        description = ''
+        if user.subreddit:
+            user_url = self.BASE_URL+user.subreddit.get('url')
+            urls = [user_url]
+            description = user.subreddit['public_description']
+            profile_urls = re.findall(r'(https?://\S+)', description)
+            profile_urls += re.findall(r'(http?://\S+)', description)
+            urls += util.trim_nulls(profile_urls)
+        else:
+            urls = [self.BASE_URL+'/user/'+username]
+
+        image = user.icon_img
+
+        return util.trim_nulls({
+          'objectType': 'person',
+          'displayName': username,
+          'image': {'url': image},
+          'id': self.tag_uri(username),
+          # numeric_id is our own custom field that always has the source's numeric
+          # user id, if available.
+          'numeric_id': user.id,
+          'published': self.timestamp_to_iso8601(user.created_utc),
+          'url': urls[0],
+          'urls': [{'value': u} for u in urls] if len(urls) > 1 else None,
+          'username': username,
+          'description': description,
+          })
+
+
+    def praw_to_object(self, thng, typ):
```

Comment:
probably want reddit terms instead of api in the method name. `post_to_object`? or do redditors really call them submissions? `submission_to_object` is fine if so. same with the rest of the methods.

Reply:
also, minor nit, `thing` and `type`? no need to save single chars 😁 and yeah `type` shadows the built-in but i do it elsewhere, no worries.

Reply:
changed thing and type :+1:  for `praw_to_object`, I named it that because that method is converting both the submission and the comment.  Figured that's good code reuse since 90% of the function would be the same.  was originally gonna name it just `something_to_object` but that was a little too vague.   and it's relying on the praw objects themselves so figured it makes sense to call it that.